File: 1 Regression/5 Support Vector Regression (SVR)/R/SVR.R
# adding dataset
dataset = read.csv('Data.csv')
dataset = dataset[2:3]

# ---------------------------------------------------
# No need to deal with missing values
# as csv file doesn't contain any column which can have missing values

# ---------------------------------------------------
# splitting a dataset into training and test
# We have only 10 rows as data so we will not split dataset

# if you haven't installed caTools library you have to run below line once.
# install.packages('caTools')

# adding lib
# library(caTools)
# 
# splitting data in test and training dataset
# set.seed(404)
# split = sample.split(dataset$Salary, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# ---------------------------------------------------
# Applying scalar
# No need to scale
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

# ---------------------------------------------------
# fitting SVR to the training set
# if you haven't installed e1071 library you have to run below line once.
# install.packages('e1071')

# adding lib
library(e1071)
regressor = svm(formula = Salary ~ .,
               data = dataset,
               type = 'eps-regression')

# Predicting the test set results
y_pred = predict(regressor, newdata = data.frame(Level = 6.5))

# ---------------------------------------------------
# if you haven't installed ggplot2 library you have to run below line once.
# install.packages('ggplot2')

# Visualising the Training set results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
            color = 'blue') +
  ggtitle('Salary vs Position (Training set)') +
  xlab('Position') +
  ylab('Salary')

File: 1.regression/5.support_vector_regression/python/svr.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# adding libs
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# adding dataset
dataset = pd.read_csv('data.csv')

# splitting a dataset into independent and dependent variables
# here 1st column is already encoded and represented as column 2 in csv file
# so we will use only 2nd column as an independent variable
X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, -1].values

# No need to deal with missing values
# as csv file doesn't contain any column which can have missing values

# No need to transform any independent variable
# as we will be using 2nd column (which is numerical) to train and test model.

# splitting a dataset into training and test
# We have only 10 rows as data so we will not split dataset

# ---------------------------------------------------
# Applying Feature Scalling

# transforming y into 2d
y = y.reshape(len(y), 1)

from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
sc_y = StandardScaler()
X = sc_X.fit_transform(X)
y = sc_y.fit_transform(y)

# ---------------------------------------------------
# Applying SVR
# RBF kernel = Radial basis function kernel
from sklearn.svm import SVR
regressor = SVR(kernel = 'rbf')
regressor.fit(X, y)

X_pred = [[6.5]]
y_pred = sc_y.inverse_transform(regressor.predict(sc_X.transform(X_pred)))

# ---------------------------------------------------
# Graph for SVR
# RED DOTS REPRESENTS THE TEST PREDICTIONS
# GREEN DOTS REPRESENTS THE TRAIN SET ACTUAL VALUES
# BLUE DOTS REPRESENTS THE REGRESSION LINE
plt.scatter(sc_X.inverse_transform(X), sc_y.inverse_transform(y), color = 'green')
plt.plot(sc_X.inverse_transform(X), sc_y.inverse_transform(regressor.predict(X)), color = 'blue')
plt.title('Position vs Salary (SVR)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
plt.show()